### PR TITLE
Redis support via Jedis client

### PR DESCRIPTION
**Pull request by gihad**
This adds Jedis 2.9.0 client support

- `spring-boot-jdbc` has been changed to `spring-boot-jdbc-redis`, redis calls were added
- Docker compose file in `spring-boot-jdbc-redis` now spins up a container with redis
- This has been running in production here for about 3 days. No noticeable overhead.
- Numbers are similar to what I get in New Relic.
- Method being instrumented is https://github.com/xetorthio/jedis/blob/master/src/main/java/redis/clients/jedis/Connection.java#L123

Some improvements that could be made:
- Sending the arguments from the redis commands to the datadog servers similar to the Ruby redis integration (note: Newrelic and Glowroot java APM also doesn't have the arguments). I haven't found a way to do this without some implications on performance.

- As seen on the last screenshot there seems to be an issue reporting really short spans

image

image

image

Sometimes when the time spent in redis is too low it's not reported accurately (e.g.: 1ns):
image

At first I thought maybe the method being instrumented wasn't doing any io which would explain the very low `1ns` recorded time for the redis span, but after looking at the Jedis code it looks like IO is included in `sendCommand`. Also, other APM tools instrument the same method I did here: e.g.: https://github.com/glowroot/glowroot/blob/master/agent/plugins/redis-plugin/src/main/resources/META-INF/glowroot.plugin.json
Finally, I see similar numbers in New Relic.

### CODE REVIEW

**Comment by tylerbenson** [comment]
Hey @gihad,
Nice work on this PR.  I wanted to see if you'd be comfortable adding some automated tests.  (Perhaps using an [embedded redis server](https://github.com/kstyrc/embedded-redis) so the real driver is exercised.)  We have a slick system that allows the tests to be run with the instrumentation automatically applied by using `AgentTestRunner`.  Please take a look at [CassandraClientTest](https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/datastax-cassandra-3.2/src/test/groovy/CassandraClientTest.groovy#L9) and [KafkaTest](https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaTest.groovy#L81-L130) for examples of the kind of verification I like to see.

If you're not comfortable with that, or you don't have more time to commit let me know.
Thanks!

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+      final Span span = scope.span();
+      Tags.DB_TYPE.set(span, "redis");
+      Tags.SPAN_KIND.set(span, Tags.SPAN_KIND_CLIENT);
+      Tags.COMPONENT.set(span, "redis-command");
+
+      span.setTag(DDTags.RESOURCE_NAME, command.name());
+      span.setTag(DDTags.SERVICE_NAME, "redis");
+      span.setTag(DDTags.SPAN_TYPE, "redis");
+      span.setTag("span.origin.type", command.getClass().getName());
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: If the command class is always the same, then this tag is probably not useful.  I added that in other instrumentation because that can be an important tool for debugging.

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    public static Scope nameResource(@Advice.Argument(0) final Command command) {
+
+      final Scope scope = GlobalTracer.get().buildSpan("redis.command").startActive(true);
+
+      final Span span = scope.span();
+      Tags.DB_TYPE.set(span, "redis");
+      Tags.SPAN_KIND.set(span, Tags.SPAN_KIND_CLIENT);
+      Tags.COMPONENT.set(span, "redis-command");
+
+      span.setTag(DDTags.RESOURCE_NAME, command.name());
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: What are some examples of `command.name()`?  Does this have the right level of cardinality to be useful?

It would be nice if we could name the resource off of more than just the command name, but perhaps the full command (assuming it doesn't change a lot).

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+import io.opentracing.util.GlobalTracer;
+import java.util.Collections;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import redis.clients.jedis.Protocol.Command;
+
+@AutoService(Instrumenter.class)
+public final class JedisInstrumentation implements Instrumenter {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: We just merged #213 which adds a new abstract parent for instrumentation to use.  Please rebase from master and change your new instrumentation to instead `extends Instrumenter.Configurable`, add a constructor that defines some useful names for this instrumentation and also add an override similar to what I did for [kafka](https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/KafkaConsumerInstrumentation.java#L45-L48) so that it is disabled by default.  This will allow folks to test it for a while if they desire.  At a later release we can then remove that and enable it by default.

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java (hunk 1 above)
Comment (gihad) [reply dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: Makes sense. Will remove.

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java (hunk 2 above)
Comment (gihad) [reply dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: Examples of commands are SET, GET, ZRANK and all the other redis commands described here: https://github.com/xetorthio/jedis/blob/2.9/src/main/java/redis/clients/jedis/Protocol.java#L244
I think even at this granularity it is useful

Having the command name + the arguments to the command instead is something I listed in the PR description as a possible improvement, however there are some things to consider:

The datadog ruby APM does that (SETEX command + arguments):

image

The NewRelic and Glowroot Java APM for redis only list the Commands like this one. I suspect it has something to do with how the Jedis code is laid out. There are 3 methods:

Method 1:
```
 protected Connection sendCommand(final Command cmd, final String... args) {
    final byte[][] bargs = new byte[args.length][];
    for (int i = 0; i < args.length; i++) {
      bargs[i] = SafeEncoder.encode(args[i]);
    }
    return sendCommand(cmd, bargs);
  }
```

Method 2:
```
protected Connection sendCommand(final Command cmd) {
    return sendCommand(cmd, EMPTY_ARGS);
  }
```

Method 3:
```
protected Connection sendCommand(final Command cmd, final byte[]... args) {
...
}
```

As you can see methods 1 and 2 are just wrappers for `method 3`. In my instrumentation I'm just monitoring method 3 and not converting the byte[] args back to String. Here are some options:

- Convert the byte[] back to String to send the arguments along the command. I decided against this for performance reasons.

- Only monitor Method 1 which has `sendCommand(final Command cmd, final String... args)`. I decided against this because we would lose instrumentation for redis commands that don't have arguments since they use method 2.

- Create separate Advices to monitor method 1 and 2 and not monitor method 3. I tried this but the code ended up convoluted and I didn't get to work. Didn't spend much time trying it tho.


Having this info do you have any suggestions on how to proceed or do you think we should just leave it the way it is (just the command name)?

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java (hunk 3 above)
Comment (gihad) [reply dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: Nice, will do.

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: I am fine having resource name be just the command name.  That is something we can iterate on in the future.  I do suggest doing what the ruby client does and setting the full command + arguments in the `redis.raw_command` tag ([ex](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/ext/redis.rb#L11)). The ruby client does have an additional level of [quantization](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/redis/quantize.rb), but I don't quite understand it so I don't expect you to implement that.

**Comment by gihad** [comment]
@tylerbenson I can address all the comments here including the embedded redis tests but I won't be able to get to it soon. Whatever time I have available for datadog work I need to spend at getting other aspects of APM working (e.g.: currently errors wrapped by spring boot don't show up in datadog).

**Comment by gihad** [comment]
@tylerbenson finally got around to working on this. All comments have been addressed.

File: dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle [hunk 4]
```diff
@@ -0,0 +1,18 @@
+
+apply from: "${rootDir}/gradle/java.gradle"
+
+dependencies {
+  compileOnly group: 'redis.clients', name: 'jedis', version: '2.9.0'
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle]: Sorry... one last thing and I think we're good.  Generally, the convention is that the version here is the earliest version the instrumentation will work with.  Will this work with anything before 2.9?

File: dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  @Override
+  protected boolean defaultEnabled() {
+    return false;
+  }
+
+  @Override
+  public AgentBuilder apply(final AgentBuilder agentBuilder) {
+    return agentBuilder
+        .type(not(isInterface()).and(hasSuperType(named("redis.clients.jedis.Connection"))))
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-2.9/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: The way we restrict the instrumentation from getting added to a version that isn't supported is by using this kind of thing: https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/aws-sdk/src/main/java/datadog/trace/instrumentation/aws/AWSClientInstrumentation.java#L34-L39.

Those classes are verified using a plugin: https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/aws-sdk/aws-sdk.gradle#L5-L14

Basically those classes are unique to the versions of the library it supports.

File: dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle (hunk 4 above)
Comment (gihad) [reply dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle]: @tylerbenson I just did a scan on all the tags available in the Jedis Repo (e.g.: https://github.com/xetorthio/jedis/blob/jedis-1.4.0/src/main/java/redis/clients/jedis/Connection.java)

My findings is that the first version that will be supported by this is Jedis `1.4.0` and it will work all the way to `2.10`. However the current class in Master has changed the method and it will no longer work in their next release so we should probably cap the support at `2.10`

Comment (gihad) [reply dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle]: Method in master has a different signature: https://github.com/xetorthio/jedis/blob/master/src/main/java/redis/clients/jedis/Connection.java#L123

Comment (tylerbenson) [reply dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle]: Since it hasn't been released yet, our tooling can't really exclude it well.  Let's just rename the project to `jedis-1.4` and add some classes to identify it differently from pre-1.4.  Would you like to do that or would you rather I do it?

Comment (gihad) [reply dd-java-agent/instrumentation/jedis-2.9/jedis-2.9.gradle]: @tylerbenson Just want to make sure you are aware that 1.4.0 is a really old version from Nov 15, 2010 and also one of the first versions (https://github.com/xetorthio/jedis/releases?after=jedis-1.5.0-RC1). Before 1.4.0 all the way down to 1.0.0 the method to instrument has this signature: https://github.com/xetorthio/jedis/blob/jedis-1.3.1/src/main/java/redis/clients/jedis/Connection.java#L52

I don't think I will get to it anytime soon so I rather you do it. Thanks.

Also, we will probably have to instrument another class after the next release of Jedis comes out (code in master now) anyways so we should plan for multiple instrumentations either way.

File: dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java [hunk 6]
```diff
@@ -40,22 +37,25 @@ protected boolean defaultEnabled() {
   @Override
   public AgentBuilder apply(final AgentBuilder agentBuilder) {
     return agentBuilder
-        .type(not(isInterface()).and(hasSuperType(named("redis.clients.jedis.Connection"))))
+        .type(
+            named("redis.clients.jedis.Protocol"),
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: I switched to the protocol instead of the connection since earlier versions of the [no-arg sendCommand](https://github.com/xetorthio/jedis/blob/jedis-1.4.0/src/main/java/redis/clients/jedis/Connection.java#L74) didn't delegate the same way.  Do you see any problems with this @gihad?

File: dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java [hunk 7]
```diff
@@ -40,22 +37,25 @@ protected boolean defaultEnabled() {
...
                     JedisAdvice.class.getName()))
         .asDecorator();
   }
 
   public static class JedisAdvice {
 
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static Scope nameResource(@Advice.Argument(0) final Command command) {
+    public static Scope startSpan(@Advice.Argument(1) final Object arg) {
+      final Command command = (Command) arg;
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: @realark I forget... is this needed for spring boot executable jar type scenarios?

**Comment by tylerbenson** [comment]
Lets get an approval from @realark then we can merge.

File: dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java (hunk 7 above)
Comment (realark) [reply dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: It's needed if the class being instrumented is in the method signature of the advice. I think Protocol is being instrumented, so I don't think this is needed.

File: dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java (hunk 6 above)
Comment (gihad) [reply dd-java-agent/instrumentation/jedis-1.4/src/main/java/datadog/trace/instrumentation/jedis/JedisInstrumentation.java]: The only thing I see is that the sendCommand method from Connection [does more work](https://github.com/xetorthio/jedis/blob/jedis-1.4.0/src/main/java/redis/clients/jedis/Connection.java#L63), not sure what's more accurate to instrument in this case. I'm good either way.

**Comment by n0mer** [comment]
@tylerbenson @gihad is Lettuce integration planned? 

spring-data-redis 2.0 supports reactive redis: https://docs.spring.io/spring-data/redis/docs/2.0.4.RELEASE/reference/html/#redis:reactive

**Comment by tylerbenson** [comment]
I’m not familiar with lettuce. I’d suggest contacting support and submit a feature request with both of those.
